src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray_1.cs

```diff
@@ -728,6 +728,9 @@ public ImmutableArray<T> Replace(T oldValue, T newValue)
         public ImmutableArray<T> Replace(T oldValue, T newValue, IEqualityComparer<T> equalityComparer)
         {
             var self = this;
+            // TODO: This is calling the IImmutableList<T> overload which is silently boxing the ImmutableArray<T>.
+            // It's also causing inconsistencies where calling Replace on an IsDefault instance will throw an
+            // InvalidOperationException rather than a NullReferenceException.
```

Comment:
I added some extra test coverage that called this method on a default `ImmutableArray`. That caused problems because the `IndexOf` call here is actually boxing this to an `IImmutableList<T>` and calling the [extension method](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableList.cs#L171) in `ImmutableList`.

The fix would be to simply add `self.ThrowNullRefIfNotInitialized()` before this line, or modify the `IndexOf` call so it calls the `ImmutableArray` overload and avoids boxing. This would be a behavioral change from `InvalidOperationException` -> `NullReferenceException` for default arrays, but I am assuming throwing IOE in the first place because normally everything throws NRE for default arrays. Is this OK to change?